Libplanet/Net/Protocols/KademliaProtocol.cs

```diff
@@ -0,0 +1,525 @@
+                {
+                    throw new Exception(
+                        "Received pong is invalid.");
+                }
+
+                if (pong.Remote.Address.Equals(_address))
+                {
+                    throw new Exception(
+                        "Cannot receive pong from self");
+                }
+
+                // update process required
+                await UpdateAsync(pong.Remote, cancellationToken);
+            }
+            catch (TimeoutException)
+            {
+                await RemovePeerAsync(target);
+                throw;
+            }
+            catch (DifferentAppProtocolVersionException)
+            {
+                _logger.Debug("Different AppProtocolVersion encountered at PingAsync.");
+            }
+            catch (Exception)
+            {
+                throw;
+            }
+        }
+
+        // This updates routing table when receiving a message.
+        // if corresponding bucket for remote peer is not full, just adds remote peer.
+        // otherwise check whether if the least recently used (LRU) peer
+        // is alive to determine evict LRU peer or discard remote peer.
+        private async Task UpdateAsync(
+            Peer peer,
+            CancellationToken cancellationToken = default(CancellationToken))
+        {
+            if (peer is null)
+            {
+                throw new ArgumentNullException(nameof(peer));
+            }
+
+            if (cancellationToken == default(CancellationToken))
+            {
+                cancellationToken = _cancellationToken;
+            }
+
+            if (cancellationToken.IsCancellationRequested)
+            {
+                throw new TaskCanceledException();
+            }
+
+            if (peer != null)
+            {
+                bool contains = _routing.Contains(peer);
+                Peer evictionCandidate = await _routing.AddPeerAsync(peer);
+                if (evictionCandidate is null)
+                {
+                    // added successfully since there was empty space in the bucket
+                    if (!contains)
+                    {
+                        _logger.Debug($"Added [{peer.Address.ToHex()}] to table");
+                    }
+                }
+                else
+                {
+                    try
+                    {
+                        await PingAsync(
+                            peer,
+                            RequestTimeout,
+                            cancellationToken);
+                        _routing.BucketOf(peer).ReplacementCache.Add(evictionCandidate);
+                    }
+                    catch (TimeoutException)
+                    {
+                        await RemovePeerAsync(peer);
+                        await _routing.AddPeerAsync(evictionCandidate);
+                    }
+                }
+            }
+
+            /*foreach (KBucket bucket in _routing.NonFullBuckets)
```

Comment:
Is this block needed?

Reply:
This block should be moved where peer is actually removed.